### PR TITLE
Prevent disclosure of user emails via Gravatar profile image URLs

### PR DESCRIPTION
**Pull request by damien**
This is a fix for the underlying issues documented in #3278.

## Summary

Users who have set their emails to be hidden have indicated they do not desire to have their email address publicly disclosed.
    
Historically, RubyGems.org has used Gravatar to provide profile images for it's users, but Gravatar is known to leak information about user emails through it's image request API that requires an md5 hash of user emails.
    
A potential fix here was to hide Gravatar profile image URLs behind a proxy, but that still leaves users open to being identified by a reverse image search of their Gravatar profile images.
    
In an effort to preserve user privacy, we will instead not reference Gravatar profile images at all for any user which has configured their account to keep their emails private.

## Notable Changes

* Remove dependency on [`gravtastic`](https://github.com/chrislloyd/gravtastic/) gem
* Add `Gravatar` model that implements a subset of the functionality previously provided by gravtastic
* Avoid referencing Gravatar images for users who have opted to keep their emails private.

## Potential Follow Up Work

* Find a nicer fall-back behavior to provide profile images for users who wish to avoid sharing their emails (and likely also tying their RubyGems.org identity to other identities they may have).
* Proxy requests to Gravatar URLs through an internal endpoint to avoid directly exposing hashed md5s of all user emails. Privacy benefit of this is marginal, as once this PR is merged the only users who would benefit from this are those who have opted to share their emails publicly on their rubygems.org profiles.

### CODE REVIEW

**Review by jenshenny — COMMENTED** [review]
Thanks for putting this up Damien! It overall looks good. 

After thinking on this a bit, the one thing I'm concerned about is that all accounts are email private by default. The intent of setting email to private would just omit their email on their profile. I'm guessing that most users will have this field and be surprised that their image is a default one after this ships. If I'm a user, I would be fine with using a gravatar and be aware of the potential leak, but not necessary let my email to be surfaced.

Perhaps we can just add a checkbox in the edit profile page to toggle whether they like a default image or their gravatar and letting them know the risk in a tooltip. wdyt?

File: Gemfile [hunk 1]
```diff
@@ -10,7 +10,6 @@ gem "clearance"
 gem "dalli"
 gem "delayed_job"
 gem "delayed_job_active_record"
-gem "gravtastic"
```
Comment (jenshenny) [inline Gemfile:13]: 🚀

**Comment by damien** [comment]
Hey @jenshenny!

> After thinking on this a bit, the one thing I'm concerned about is that all accounts are email private by default. The intent of setting email to private would just omit their email on their profile. I'm guessing that most users will have this field and be surprised that their image is a default one after this ships. If I'm a user, I would be fine with using a gravatar and be aware of the potential leak, but not necessary let my email to be surfaced.
> 
> Perhaps we can just add a checkbox in the edit profile page to toggle whether they like a default image or their gravatar and letting them know the risk in a tooltip. wdyt?

I agree with you, but regardless of the path we take here an end user will need to take action. I think what you're getting at here is what action a user needs to take and what happens if they do or do not take action.

In matters of privacy, it's generally ethical and safer to err on the side of caution and require end-users to opt-in to anything that exposes their information to the public at large. It's hard for me to really gauge how big of a problem or risk this is for individual users of RubyGems.org in practice, but if we're not sure I'd advocate for playing it safe—especially if the only downside is users needing to re-enable Gravatar for their profiles the next time they log in. Historically, bad things have happened to people who had their privacy compromised if they happened to live somewhere where their work was objectionable by their local government.

That said, leaving the current behavior maintains the status-quo, which as I understand it is:

1. Existing users maintain the same level of risk exposure around email leaks
2. New users are put at risk of their email being exposed via Gravatar

The ship has sailed for existing users—nothing we do here changes any email exposures that could have happened in the past. I think the more problematic thing here is that leaving things as they are opens new users up to email exposure without their consent or awareness.

The opt-in checkbox approach seems very workable, though it'd need to get added to both a user's settings and the sign up form to avoid scenarios where users risk email exposure through usage of Gravatar without opting in.

We could probably also let existing users default to using Gravatar, but if we go that route I'd strongly recommend sending an announcement out to all RubyGems.org account holders prompting them to opt-out if they're concerned about potential email leaks—this ensures people who do have strong privacy needs have the opportunity to respond appropriately in the context of their own privacy needs.

---

As a closing thought, it may be worth considering what RubyGems.org's policy is around privacy in relation to other features or priorities. If maintaining privacy is something important to the service and org, that's something that should dictate decisions around privacy defaults, how it sets expectations around user privacy, and what considerations should be made for future features, fixes, or initiatives.

It's OK for this to not be considered a privacy-preserving service if the burden of that is too high, but that needs to be obvious and easily understood to all users if that is the reality.

**Comment by jenshenny** [comment]
Appreciate all your thoughts around this 🙌 

> The opt-in checkbox approach seems very workable, though it'd need to get added to both a user's settings and the sign up form to avoid scenarios where users risk email exposure through usage of Gravatar without opting in.

Oh yes, good catch. I agree the best practice would be to require end-users to opt-in to anything that exposes their information, and we can set that precedent for new users. As a followup it can decided to do the same for existing users and let them explicitly re-enable gravatar if they wish, or implement a proxy for the URLs. 

You mentioned that implementing a proxy still leaves users open to being identified by a reverse image search of their Gravatar profile images. If users are concerned about being identified by a reverse image search, they can disable this new checkbox and not use Gravatar (I think users should know the risk of reverse image search when they first upload their picture onto Gravatar).

**To summarize my thoughts, right now we can:**
Add checkbox to enable/disable their gravatar image in sign up and edit profile pages. New users would need to opt into enabling Gravatar.

**In the future:**
Add a proxy to hide md5 hashes / require existing users to opt into Gravatar again after describing the risk.

I'm going to tag @simi and @indirect in case they have thoughts on how this should go about and RubyGems.org's thoughts around privacy.

File: app/models/gravatar.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
+class Gravatar
+  GRAVATAR_ENDPOINT = 'https://www.gravatar.com/avatar/'.freeze
+
+  # A bogus Gravatar ID we can use to avoid leaking the email hash
+  # of users that prefer to not have their emails disclosed.
+  #
+  # @see https://en.gravatar.com/site/implement/images/#default-image
+  GRAVATAR_DEFAULT_ID = "00000000000000000000000000000000"
```
Comment (jenshenny) [inline app/models/gravatar.rb:8]: Right now this will mean that everyone would have the same default image. Could we default it to perhaps `user_id % 5` or another operation/number so not all users will have the same image? 

eg. https://www.gravatar.com/avatar/1?d=retro would be different from https://www.gravatar.com/avatar/0?d=retro

File: app/models/gravatar.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
+class Gravatar
```
Comment (colby-swandale) [inline app/models/gravatar.rb:1]: Would this fit better as a [Rails Concern](https://api.rubyonrails.org/v7.0.4/classes/ActiveSupport/Concern.html) instead? I feel this would better encapsulate this functionality and remove the depends on a `User` object. May also even reduce `app/helpers/application_helper.rb` I think?

**Comment by segiddins** [comment]
@simi can we close this out now we've merged the other PR?

**Comment by simi** [comment]
> @simi can we close this out now we've merged the other PR?

Yes, this is fixed.
